File: app/src/main/java/com/simplecity/amp_library/utils/menu/album/AlbumMenuUtils.java
package com.simplecity.amp_library.utils.menu.album;

import android.content.Context;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.Toolbar;
import android.view.SubMenu;
import com.simplecity.amp_library.R;
import com.simplecity.amp_library.model.Album;
import com.simplecity.amp_library.model.Song;
import com.simplecity.amp_library.playback.MediaManager;
import com.simplecity.amp_library.utils.MusicUtils;
import com.simplecity.amp_library.utils.PlaylistUtils;
import com.simplecity.amp_library.utils.ShuttleUtils;
import com.simplecity.amp_library.utils.extensions.AlbumExt;
import com.simplecity.amp_library.utils.menu.MenuUtils;
import io.reactivex.Single;
import java.util.List;

public class AlbumMenuUtils {

    public interface Callbacks {

        void onPlaylistItemsInserted();

        void onQueueItemsInserted(String message);

        void playNext(Single<List<Song>> songsSingle);

        void showTagEditor(Album album);

        void showDeleteDialog(Album album);

        void showDeleteDialog(List<Album> albums);

        void showDeleteDialog(Single<List<Album>> albums);

        void showAlbumInfo(Album album);

        void showArtworkChooser(Album album);

        void showUpgradeDialog();

        void showToast(String message);
    }

    private AlbumMenuUtils() {
        //no instance
    }

    public static void setupAlbumMenu(PopupMenu menu) {
        menu.inflate(R.menu.menu_album);

        // Add playlist menu
        SubMenu sub = menu.getMenu().findItem(R.id.addToPlaylist).getSubMenu();
        PlaylistUtils.createPlaylistMenu(sub);
    }

    public static Toolbar.OnMenuItemClickListener getAlbumMenuClickListener(Context context, MediaManager mediaManager, Single<List<Album>> selectedAlbums, Callbacks callbacks) {
        return item -> {
            switch (item.getItemId()) {
                case MediaManager.NEW_PLAYLIST:
                    MenuUtils.newPlaylist(context, AlbumExt.INSTANCE.getSongsForAlbums(selectedAlbums), callbacks::onPlaylistItemsInserted);
                    return true;
                case MediaManager.PLAYLIST_SELECTED:
                    MenuUtils.addToPlaylist(context, item, AlbumExt.INSTANCE.getSongsForAlbums(selectedAlbums), callbacks::onPlaylistItemsInserted);
                    return true;
                case R.id.playNext:
                    callbacks.playNext(AlbumExt.INSTANCE.getSongsForAlbums(selectedAlbums));
                    return true;
                case R.id.addToQueue:
                    MenuUtils.addToQueue(mediaManager, AlbumExt.INSTANCE.getSongsForAlbums(selectedAlbums), callbacks::onQueueItemsInserted);
                    return true;
                case R.id.delete:
                    callbacks.showDeleteDialog(selectedAlbums);
                    return true;
            }
            return false;
        };
    }

    public static PopupMenu.OnMenuItemClickListener getAlbumMenuClickListener(Context context, MediaManager mediaManager, Album album, Callbacks callbacks) {
        return item -> {
            switch (item.getItemId()) {
                case R.id.play:
                    MenuUtils.play(mediaManager, AlbumExt.INSTANCE.getSongsForAlbum(album), callbacks::showToast);
                    return true;
                case R.id.playNext:
                    callbacks.playNext(AlbumExt.INSTANCE.getSongsForAlbum(album));
                    return true;
                case MediaManager.NEW_PLAYLIST:
                    MenuUtils.newPlaylist(context, AlbumExt.INSTANCE.getSongsForAlbum(album), callbacks::onPlaylistItemsInserted);
                    return true;
                case MediaManager.PLAYLIST_SELECTED:
                    MenuUtils.addToPlaylist(context, item, AlbumExt.INSTANCE.getSongsForAlbum(album), callbacks::onPlaylistItemsInserted);
                    return true;
                case R.id.addToQueue:
                    MenuUtils.addToQueue(mediaManager, AlbumExt.INSTANCE.getSongsForAlbum(album), callbacks::onQueueItemsInserted);
                    return true;
                case R.id.editTags:
                    if (ShuttleUtils.isUpgraded()) {
                        callbacks.showTagEditor(album);
                    } else {
                        callbacks.showUpgradeDialog();
                    }
                    return true;
                case R.id.info:
                    callbacks.showAlbumInfo(album);
                    return true;
                case R.id.artwork:
                    callbacks.showArtworkChooser(album);
                    return true;
                case R.id.blacklist:
                    MenuUtils.blacklist(AlbumExt.INSTANCE.getSongsForAlbum(album));
                    return true;
                case R.id.delete:
                    callbacks.showDeleteDialog(album);
                    return true;
            }
            return false;
        };
    }
}

File: app/src/main/java/com/simplecity/amp_library/utils/menu/album/AlbumMenuUtils.kt
package com.simplecity.amp_library.utils.menu.album

import android.support.v7.widget.PopupMenu
import android.support.v7.widget.Toolbar
import com.simplecity.amp_library.R
import com.simplecity.amp_library.model.Album
import com.simplecity.amp_library.model.Playlist
import com.simplecity.amp_library.playback.MediaManager.Defs
import com.simplecity.amp_library.utils.playlists.PlaylistManager
import com.simplecity.amp_library.utils.playlists.PlaylistMenuHelper
import io.reactivex.Single

object AlbumMenuUtils {

    const val TAG = "AlbumMenuUtils"

    fun setupAlbumMenu(menu: PopupMenu, playlistMenuHelper: PlaylistMenuHelper, showGoToArtist: Boolean = true) {
        menu.inflate(R.menu.menu_album)

        if (!showGoToArtist) {
            menu.menu.findItem(R.id.go_to).isVisible = false
        }

        // Add playlist menu
        val subMenu = menu.menu.findItem(R.id.addToPlaylist).subMenu
        playlistMenuHelper.createPlaylistMenu(subMenu)
    }

    fun getAlbumMenuClickListener(selectedAlbums: Single<List<Album>>, callbacks: AlbumsMenuCallbacks): Toolbar.OnMenuItemClickListener {
        return Toolbar.OnMenuItemClickListener { item ->
            when (item.itemId) {
                Defs.NEW_PLAYLIST -> {
                    callbacks.createPlaylistFromAlbums(selectedAlbums)
                    return@OnMenuItemClickListener true
                }
                Defs.PLAYLIST_SELECTED -> {
                    callbacks.addAlbumsToPlaylist(item.intent.getSerializableExtra(PlaylistManager.ARG_PLAYLIST) as Playlist, selectedAlbums)
                    return@OnMenuItemClickListener true
                }
                R.id.playNext -> {
                    callbacks.playAlbumsNext(selectedAlbums)
                    return@OnMenuItemClickListener true
                }
                R.id.addToQueue -> {
                    callbacks.addAlbumsToQueue(selectedAlbums)
                    return@OnMenuItemClickListener true
                }
                R.id.delete -> {
                    callbacks.deleteAlbums(selectedAlbums)
                    return@OnMenuItemClickListener true
                }
            }
            false
        }
    }

    fun getAlbumMenuClickListener(album: Album, callbacks: AlbumsMenuCallbacks): PopupMenu.OnMenuItemClickListener {
        return PopupMenu.OnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.play -> {
                    callbacks.play(album)
                    return@OnMenuItemClickListener true
                }
                R.id.playNext -> {
                    callbacks.playAlbumsNext(album)
                    return@OnMenuItemClickListener true
                }
                Defs.NEW_PLAYLIST -> {
                    callbacks.createPlaylistFromAlbums(album)
                    return@OnMenuItemClickListener true
                }
                Defs.PLAYLIST_SELECTED -> {
                    callbacks.addAlbumsToPlaylist(item.intent.getSerializableExtra(PlaylistManager.ARG_PLAYLIST) as Playlist, album)
                    return@OnMenuItemClickListener true
                }
                R.id.addToQueue -> {
                    callbacks.addAlbumsToQueue(album)
                    return@OnMenuItemClickListener true
                }
                R.id.editTags -> {
                    callbacks.editTags(album)
                    return@OnMenuItemClickListener true
                }
                R.id.info -> {
                    callbacks.albumInfo(album)
                    return@OnMenuItemClickListener true
                }
                R.id.artwork -> {
                    callbacks.editArtwork(album)
                    return@OnMenuItemClickListener true
                }
                R.id.blacklist -> {
                    callbacks.blacklistAlbums(album)
                    return@OnMenuItemClickListener true
                }
                R.id.delete -> {
                    callbacks.deleteAlbums(album)
                    return@OnMenuItemClickListener true
                }
                R.id.goToArtist -> {
                    callbacks.goToArtist(album)
                    return@OnMenuItemClickListener true
                }
            }
            false
        }
    }
}